neqo-transport/src/connection/mod.rs

```diff
@@ -594,9 +594,25 @@ impl Connection {
     fn make_resumption_token(&mut self) -> ResumptionToken {
         debug_assert_eq!(self.role, Role::Client);
         debug_assert!(self.crypto.has_resumption_token());
+        // Values less than GRANULARITY are ignored when using the token, so use 0 where needed.
         let rtt = self.paths.primary().map_or_else(
-            || RttEstimate::default().estimate(),
-            |p| p.borrow().rtt().estimate(),
+            // If we don't have a path, we don't have an RTT.
+            || Duration::from_millis(0),
+            |p| {
+                let rtt = p.borrow().rtt().estimate();
+                if self.stats.borrow().frame_rx.ack == 0 {
```

Comment:
I consider `stats` a purely auxiliary mechanism, used for telemetry only. Is that consideration wrong?

If not, I wouldn't consider depending on `stats` here clean.

How about using `RttEstimate::first_sample_time` instead?

https://github.com/mozilla/neqo/blob/75372c214c5a5d1d220e5cc502afffad8d820e21/neqo-transport/src/rtt.rs#L34

E.g. something along the lines of:

```suggestion
                if p.borrow().rtt().first_sample_time().is_none() {
```

Reply:
Because we set `first_sample_time` to something even if that something is just an RTT guess and not a measured RTT. (I used that first, thinking exactly the same, but it doesn't work.)

Reply:
`RttEstimate::update` takes a `bool` `confirmed`:

https://github.com/mozilla/neqo/blob/75372c214c5a5d1d220e5cc502afffad8d820e21/neqo-transport/src/rtt.rs#L86-L93

`RttEstimate::update` is called by two functions:

- https://github.com/mozilla/neqo/blob/75372c214c5a5d1d220e5cc502afffad8d820e21/neqo-transport/src/recovery/mod.rs#L559-L565
- https://github.com/mozilla/neqo/blob/75372c214c5a5d1d220e5cc502afffad8d820e21/neqo-transport/src/path.rs#L970

How about replacing `confirmed with additional state:

``` rust
struct RttEstimate {
  // ....
  best_source: Source,
}

pub enum Source {
  Estimate,
  Ack,
  /// Ack on a confirmed connection.
  AckConfirmed,
}

impl RttEstimate {
    pub fn update(
        &mut self,
        qlog: &NeqoQlog,
        mut rtt_sample: Duration,
        ack_delay: Duration,
        // Provide the root source of the update.
        source: Source,
        now: Instant,
    ) {

```

`rtt_sample` could then call `RttEstimate::update` with `Source::Ack` or `Source::AckConfirmed`. `discard_packet` would pass `Source::Estimate`.

`RttEstimate` can then expose `best_source` to e.g. `Connection` above, or expose `RttEstimate::estimate` and `RttEstimate::guesstimate` based on `Source`.

What do you think?